nomic/data_operations.py

```diff
@@ -1,42 +1,602 @@
+        return label_to_datums
+
+    def tag(self, ids: List[str], tags: List[str]):
+        '''
+        Adds tags to datapoints.
+
+        Args:
+            ids: The datum ids you want to tag
+            tags: A list containing the tags you want to apply to these data points.
+
+        '''
+        assert isinstance(ids, list), 'ids must be a list of strings'
+        assert isinstance(tags, list), 'tags must be a list of strings'
+
+        colname = json.dumps(
+            {
+                'project_id': self.project.id,
+                'atlas_index_id': self.projection.atlas_index_id,
+                'type': 'datum_id',
+                'tags': tags,
+            }
+        )
+        payload_table = pa.table([pa.array(ids, type=pa.string())], [colname])
+        buffer = io.BytesIO()
+        writer = ipc.new_file(buffer, payload_table.schema, options=ipc.IpcWriteOptions(compression='zstd'))
+        writer.write_table(payload_table)
+        writer.close()
+        payload = buffer.getvalue()
+
+        headers = self.project.header.copy()
+        headers['Content-Type'] = 'application/octet-stream'
+        response = requests.post(self.project.atlas_api_path + "/v1/project/tag/add", headers=headers, data=payload)
+        if response.status_code != 200:
+            raise Exception("Failed to add tags")
+
+    def remove_tags(self, ids: List[str], tags: List[str], delete_all: bool = False) -> bool:
+        '''
+        Deletes the specified tags from the given data points.
+
+        Args:
+            ids: The datum_ids to delete tags from.
+            tags: The list of tags to delete from the data points. Each tag will be applied to all data points in `ids`.
+            delete_all: If true, ignores datum_ids and deletes all specified tags from all data points.
+
+        Returns:
+            True on success
+
+        '''
+        assert isinstance(ids, list), 'datum_ids must be a list of strings'
+        assert isinstance(tags, list), 'tags must be a list of strings'
+
+        colname = json.dumps(
+            {
+                'project_id': self.project.id,
+                'atlas_index_id': self.projection.atlas_index_id,
+                'type': 'datum_id',
+                'tags': tags,
+                'delete_all': delete_all,
+            }
+        )
+        payload_table = pa.table([pa.array(ids, type=pa.string())], [colname])
+        buffer = io.BytesIO()
+        writer = ipc.new_file(buffer, payload_table.schema, options=ipc.IpcWriteOptions(compression='zstd'))
+        writer.write_table(payload_table)
+        writer.close()
+        payload = buffer.getvalue()
+
+        headers = self.project.header.copy()
+        headers['Content-Type'] = 'application/octet-stream'
+        response = requests.post(self.project.atlas_api_path + "/v1/project/tag/delete", headers=headers, data=payload)
+        if response.status_code != 200:
+            raise Exception("Failed to delete tags")
```

Comment:
do we want a `__repr__` method to stay consistent with the other classes?

Reply:
What should this be. The issue with tags is that we don't get global access of all tag state

Reply:
I will add a .df method for consistency

Reply:
I have added a .df method which contains as columns a multi-hot vector where ones exist when the associated row has that tag.

Reply:
yeah good point, resolving